File: A_Star/amtrak.py
import sys
import math

def reconstruct_path(cameFrom, current):
    path = [current]
    while current in cameFrom.keys():
        current = cameFrom[current]
        path.append(current)
    return list(reversed(path))

def a_star(start, goal, graph, heuristic):
    '''
    start: starting city
    goal: ending city
    graph: currently edge list in list
    heuristic: 2-layer dictionary h[start][end] get heuristic from start to end
    Returns a list containing the path from start to end or None if no
    path exists
    '''
    all_cities = set()
    for c1, c2, weight in edge_list:
        all_cities.add(c1)
        all_cities.add(c2)
    if start not in all_cities or goal not in all_cities:
        return None, 0
    if start == goal:
        return [start], 0
    openSet = {start}
    cameFrom = dict()
    gScore = {city: math.inf for city in all_cities}
    gScore[start] = 0
    fScore = {city: math.inf for city in all_cities}
    fScore[start] = 0

    while openSet:
        current = min([(fScore[node], node) for node in openSet])[1]
        if current == goal:
            return reconstruct_path(cameFrom, current), gScore[current]
        openSet.remove(current)
        for cur, adj, cost in edge_list:
            # found an neighbor edge
            if cur == current or adj == current:
                # make sure cur is the current node and adj is the neighbor node
                if adj == current:
                    adj, cur = cur, adj
                # get a tenative score for the edge
                tenative = gScore[current] + cost
                # add/replace the neighbors values if we found a shorter path
                if tenative < gScore[adj]:
                    cameFrom[adj] = current
                    gScore[adj] = tenative
                    fScore[adj] = gScore[adj] + heuristic[adj][goal]
                    openSet.add(adj)
    return None, 0

if __name__ == '__main__':
    # Set up data structures
    graph_file = 'routes.txt'
    heur_file = 'euclidian.txt'
    with open(graph_file, 'r') as f:
        edge_list = []
        for line in f:
            # (Portland, Seattle, 174.3)\n 
            u, v, w = line.strip()[1:-1].split(',')
            # ["Portland", " Seattle", " 174.3"]
            edge_list.append( (u.strip(), v.strip(), float(w)) )

    with open(heur_file, 'r') as f:
        # heuristics is a 2-layer dictionary
        # first layer gets a dictionary of heuristics for the given
        # source city. second layer takes the goal city as a key
        # and retireves the euclidean distance from start to goal
        heuristics = dict()
        for line in f:
            start, end, val = line.strip().split(' ')
            val = float(val)
            if start not in heuristics:
                heuristics[start] = {}
            heuristics[start][end] = val
    start = sys.argv[1]
    goal = sys.argv[2]
    path, distance = a_star(start, goal, edge_list, heuristics)
    if path == None:
        print("Path does not exist between", start, "and", goal)
    else:
        print(' -> '.join(path))
        print("Total distance traveled:", distance)

File: A_Star/C++/amtrak.cpp
#include <iostream>
#include <string>
#include <cstring>
#include <unordered_map>
#include <vector>
#include <unordered_set>
#include <fstream>

#define NEIGHBOR g->edges[i].dest

using namespace std;

const float INFTY = 99999.0;

struct Edge {
  Edge();
  Edge(char*, char*, float);
  string src;
  string dest;
  float weight;
};

struct Graph {
  Graph();
  int nvertices;
  vector<Edge> edges;
  unordered_set<string> cities;
  unordered_map<string, unordered_map<string, float>> heuristic;
};

Graph::Graph() {
  nvertices = 0;
}

Edge::Edge() {}
Edge::Edge(char* s, char* d, float w) {
  src = s;
  dest = d;
  weight = w;
}

void set_heuristic(Graph *g) {
    char src[100], dst[100];
    float wt;
    ifstream f;
    f.open("euclidian.txt");
    
    f.get(src, 100, ' ');
    f.ignore(100, ' ');
    while(!f.eof()) {
        f.get(dst, 100, ' ');
        f >> wt;
        f.ignore(100, '\n');
        g->heuristic[src][dst] = wt;
        f.get(src, 100, ' ');
        f.ignore(100, ' ');
    }
    f.close();
}

Graph* process_file(void) {
  string fname = "routes.txt";
  Graph * g = new Graph;
  ifstream f;
  char bufsrc[100], bufdst[100];
  float weight;
  f.open(fname);
  while(!f.eof()) {
    f.ignore(100, '(');
    f.get(bufsrc, 100, ',');
    if(f.eof())
      break;
    f.ignore(100, ' ');
    f.get(bufdst, 100, ',');
    f.ignore();
    f >> weight;
    f.ignore(100, '\n');
    g->edges.push_back(Edge(bufsrc, bufdst, weight));
    g->edges.push_back(Edge(bufdst, bufsrc, weight));
    g->cities.insert(bufsrc);
    g->cities.insert(bufdst);
  }
  f.close();
  g->nvertices = g->cities.size();
  set_heuristic(g);
  return g;
}

string get_min(unordered_set<string> *openSet, unordered_map<string, float> *scores) {
  float min = INFTY;
  string current_low;

  for(auto& city: *openSet) {
    if((*scores)[city] < min) {
      min = (*scores)[city];
      current_low = city;
    }
  }
  openSet->erase(current_low);
  return current_low;
}

void make_path(unordered_map<string, string> cameFrom, string current)
{
  vector<string> path;
  path.push_back(current);
  while(cameFrom.find(current) != cameFrom.end()) {
    current = cameFrom[current];
    path.push_back(current);
  }
  for(int i = path.size() - 1; i >= 1; --i) {
    cout << path[i] << " --> ";
  }
  cout << path[0] << endl;
}

float a_star(Graph *g, string start, string end) {
  int n = g->edges.size();
  unordered_map<string, string> cameFrom;
  unordered_map<string, float> gscore, fscore;
  unordered_set<string> openSet;
  string current;
  float tenative;
  // city doesnt exist, just end
  if(!g->cities.count(start) || !g->cities.count(end))
    return -1.0;
  // Set up scoring maps
  for(const auto& elem:g->cities) {
    gscore[elem] = INFTY;
    fscore[elem] = INFTY;
  }
  gscore[start] = 0;
  fscore[start] = 0;

  openSet.insert(start);

  while(!openSet.empty()) {
    current = get_min(&openSet, &fscore);
    if(current == end) {
      make_path(cameFrom, current);
      return gscore[end];
    }
    // adjust neighbors score
    for(int i = 0; i < n; ++i) {
      if(g->edges[i].src != current)
        continue;
      tenative = gscore[current] + g->edges[i].weight;
      if(tenative < gscore[NEIGHBOR]) {
        cameFrom[NEIGHBOR] = current;
        gscore[NEIGHBOR] = tenative;
        fscore[NEIGHBOR] = gscore[NEIGHBOR]
            + g->heuristic[current][NEIGHBOR];
        openSet.insert(NEIGHBOR);
      }
    }
  }

  return 0.0;
}


int main(int argc, char** argv)
{
  if(argc < 3) {
    cout << "Usage: ./amtrak <start> <end>" << endl;
    return 0;
  }
  string start(argv[1]);
  string goal(argv[2]);
  Graph *g = process_file();
  float distance = a_star(g, start, goal);
  cout << "Total distance traveled: " << distance << "." << endl;

  delete g;
  return 0;
}
